android/pytorch_android/src/main/cpp/pytorch_jni.cpp

```diff
@@ -0,0 +1,683 @@
+  } else if (kTensorTypeCodeInt32 == typeCode) {
+    inputDataElementSizeBytes = 4;
+    inputTypeMeta = caffe2::TypeMeta::Make<int>();
+  } else if (kTensorTypeCodeByte == typeCode) {
+    inputDataElementSizeBytes = 1;
+    inputTypeMeta = caffe2::TypeMeta::Make<uint8_t>();
+  } else {
+    facebook::jni::throwNewJavaException(
+        facebook::jni::gJavaLangIllegalArgumentException, "Unknown typeCode");
+  }
+
+  const auto inputDataCapacity =
+      directBufferCapacity * inputDataElementSizeBytes;
+  if (inputDataCapacity != (inputNumel * inputDataElementSizeBytes)) {
+    facebook::jni::throwNewJavaException(
+        facebook::jni::gJavaLangIllegalArgumentException,
+        "Tensor dimensions(elements number:%d, element byte size:%d, total "
+        "bytes:%d) inconsistent with buffer capacity(%d)",
+        inputNumel,
+        inputDataElementSizeBytes,
+        inputNumel * inputDataElementSizeBytes,
+        inputDataCapacity);
+  }
+
+  at::Tensor inputTensor = torch::empty(torch::IntArrayRef(inputDimsVec));
+  inputTensor.unsafeGetTensorImpl()->ShareExternalPointer(
+      {jni->GetDirectBufferAddress(inputData.get()), at::DeviceType::CPU},
+      inputTypeMeta,
+      inputDataCapacity);
+  return inputTensor;
+}
+
+class JTensor : public facebook::jni::JavaClass<JTensor> {
+ public:
+  constexpr static const char* kJavaDescriptor = "Lorg/pytorch/Tensor;";
+
+  static facebook::jni::local_ref<JTensor> newJTensor(
+      facebook::jni::alias_ref<facebook::jni::JByteBuffer> jBuffer,
+      facebook::jni::alias_ref<jintArray> jDims,
+      jint typeCode) {
+    static auto jMethodNewTensor =
+        JTensor::javaClassStatic()
+            ->getStaticMethod<facebook::jni::local_ref<JTensor>(
+                facebook::jni::alias_ref<facebook::jni::JByteBuffer>,
+                facebook::jni::alias_ref<jintArray>,
+                jint)>("nativeNewTensor");
+    return jMethodNewTensor(
+        JTensor::javaClassStatic(), jBuffer, jDims, typeCode);
+  }
+
+  static facebook::jni::local_ref<JTensor> newJTensorFromAtTensor(
+      const at::Tensor& tensor) {
+    const auto scalarType = tensor.scalar_type();
+    int typeCode = 0;
+    if (at::kFloat == scalarType) {
+      typeCode = kTensorTypeCodeFloat32;
+    } else if (at::kInt == scalarType) {
+      typeCode = kTensorTypeCodeInt32;
+    } else if (at::kByte == scalarType) {
+      typeCode = kTensorTypeCodeByte;
+    } else {
+      facebook::jni::throwNewJavaException(
+          facebook::jni::gJavaLangIllegalArgumentException,
+          "at::Tensor scalar type is not supported on java side");
+    }
+
+    const auto& tensorDims = tensor.sizes();
+    std::vector<int> tensorDimsVec;
+    for (const auto& dim : tensorDims) {
+      tensorDimsVec.push_back(dim);
+    }
+
+    facebook::jni::local_ref<jintArray> jTensorDims =
+        facebook::jni::make_int_array(tensorDimsVec.size());
+
+    jTensorDims->setRegion(0, tensorDimsVec.size(), tensorDimsVec.data());
```

Comment:
I know caffe2 uses `int[]` for dims, but maybe we should switch to `long[]`?  Then we could just do:

```
tensorDims = tensor.sizes();
jTensorDims = make_long_array(tensorDims.size());
jTensorDims->setRegion(0, tensorDims.size(), tensorDims.data());
```

Could also consider adding an overload to `make_long_array` that takes a container like:

```
template<typename ContiguousContainer>
inline local_ref<jlongArray> make_long_array(const RandomAccessContainer& container) {
  ContiguousContainer::iterator b = std::begin(container);
  ContiguousContainer::iterator e = std::end(container);
  ContiguousContainer::difference_type size = e - b;
  local_ref<jlongArray> arr = make_long_array(size);
  arr->setRegion(0, size, b);
  return arr;
}
```

Then, this code just becomes

```
local_ref<jintArray> = make_long_array(tensor.sizes());
```

Doesn't have to be done in this diff, though.

Reply:
Agree, at::Tensor uses int64_t for sizes() (dims) - I will change dims to long[]